HRQB 11 - Data Warehouse client and SQLQueryExtractTask base task

ghukill opened the pull request:
### Purpose and background context

This PR introduces class `DWClient` to assist with making queries to the Oracle Data Warehouse.  The goal was to keep this as simple as possible, assisting with setting up a connection, supporting testing to a local SQLite database, and returning the results of a query as a pandas DataFrame.

Additionally, a new base task class was created `SQLQueryExtractTask`.  This base class is designed to assist with Extract tasks that query for data from the warehouse, using the `DWClient`, and return a DataFrame for downstream tasks to work with.  This leans into the requirement of `PandasPickleTask` to define a `get_dataframe()` method, which in turn this defines as an opinionated method to perform a SQL query.

The end result is a task that can look as simple as this:

```python
class ExtractTable1AndTable2Data(SQLQueryExtractTask):
    pipeline = luigi.Parameter()
    stage = luigi.Parameter("Extract")

    @property
    def sql_query(self) -> str:
        return """
        select
           t1.column_a,
           t1.column_b,
           t2.column_c
        from my_table t1
        inner join my_other_table t2 on t1.column_a = t2.column_a
        where t2.column_d = 'foo'
        """
```

When this task runs, it will use the `sql_query` property defined, perform a query, and return the result as a DataFrame.  The goal is to focus on writing the SQL queries themselves, and not the orchestration of executing queries or preparing the results for downstream tasks.

### How can a reviewer manually see the effects of these changes?

An ipython shell is likely the easiest way to get a feel for the new base task type `SQLQueryExtractTask` and `DWClient` working together.

Set env vars:
```
WORKSPACE=dev
SENTRY_DSN=None
DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH
LUIGI_CONFIG_PATH=hrqb/luigi.cfg
QUICKBASE_API_TOKEN=abc123
QUICKBASE_APP_ID=def456
```

Ipython shell:
```python
# import a testing Task class from fixtures
from tests.fixtures.tasks.extract import SQLExtractAnimalColors

# instantiate the Task with pipeline name and stage
task = SQLExtractAnimalColors(pipeline="Testing", stage="Extract")

# run get_dataframe() method
# this relies on the base class SQLQueryExtractTask to perform a SQL query
# (configured here to query a local SQLite database that has data)
# and returns a dataframe only, no files are written
task.get_dataframe()
# Out[3]: 
#   animal_id  color
# 0         42  green
# 1        101    red

# run actual run() method, and note the file created:
# output/Testing__Extract__SQLExtractAnimalColors.pickle
task.run()

# other confirmations that data was written identical
# to the PandasPickleTask this task class extends

task.path
# Out[4]: 'output/Testing__Extract__SQLExtractAnimalColors.pickle'

task.target
# Out[5]: <hrqb.base.target.PandasPickleTarget at 0x12de27e10>

task.target.exists()
# Out[6]: True

task.complete()
# Out[7]: True
```

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
YES; hypothetical connection to data warehouse, though credentials are not deployed yet

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/HRQB-11

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

ghukill committed:
Add DWClient for Data Warehouse work

Why these changes are being introduced:

A primary source of data for this application to work with
is data coming from the MIT Oracle Data Warehouse.  A normalized
way of connecting and performing queries is needed.

How this addresses that need:
* Creates new DWClient to connect to, and query from, a remote
database.
* DWClient includes a method execute_query that uses pandas
to return a DataFrame of the results, leaning into the
DataFrame-first nature of the PandasPickleTasks.

Side effects of this change:
* Connection possible to external Data Warehouse

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/HRQB-11

ghukill committed:
Test utils functions

ghukill committed:
Use factory pattern for attrs default

Why these changes are being introduced:
For testing purposes it can be difficult to test when
an attrs field gets a default from an env var at class
definition, as this will not allow monkeypatching.

How this addresses that need:
* Use factory pattern where the default value is set
each instantiation of the class via a lambda

Side effects of this change:
* None

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/HRQB-13

ghukill committed:
Add SQLQueryExtractTask base task

Why these changes are being introduced:

Similar to base task QuickbaseUpsertTask, a very common extract task
will be querying the Oracle Data Warehouse for data.  To reduce boilerplate
code in each of these defined tasks, a base class will be helpful for
setting up some structure to follow.

How this addresses that need:
* New base task SQLQueryExtractTask
  * requires only a sql_query property, where the rest of querying and
writing pandas dataframe as task target is handled
  * optional SQL query parameters may be defined
  * optional DWClient instance may be defined (e.g. testing to
SQLite database)

Side effects of this change:
* None

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/HRQB-11

ghukill committed:
sqlalchemy dependency and linting updates

ghukill committed:
update README with new env var

ghukill commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,62 @@
+    """Client to provide Oracle Data Warehouse connection and querying."""
+
+    connection_string: str = field(
+        factory=lambda: Config().DATA_WAREHOUSE_CONNECTION_STRING,
```
Unsure if I commented on this in tests, but using a `factory=lambda: ...` pattern here is helpful for testing.  This ensures that only when the `DWClient` is instantiated, does it pull the env var `DATA_WAREHOUSE_CONNECTION_STRING` if an explicit connection string is not passed.

If this had been `default=Config().DATA_WAREHOUSE_CONNECTION_STRING`, then monkey patching the env var at pytest time would not be sufficient, as this class is already defined with the default value for any future instances being the env var _before_ that monkey patching.

jonavellecuerdo replied:
[Non-blocking] Maybe pull some of this information into the doc string or as a comment above the declaration of `connection_string`?

ghukill committed:
SQLQueryExtractTask read SQL files

Why these changes are being introduced:

Some extract tasks that query the data warehouse will
have complex SQL queries.  Storing them in a dedicated
file will support syntax highlighting and testing of
those files directly, while keeping the task definitions
sipmler.

How this addresses that need:
* Add new property SQLQueryExtractTask.sql_file
* Required either sql_query OR sql_file defined

Side effects of this change:
* None

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/HRQB-11

ehanson8 commented on hrqb/base/task.py:
```diff
@@ -125,6 +126,45 @@ def run(self) -> None:
+        """Optional parameters to include with SQL query."""
+        return {}
+
+    def get_dataframe(self) -> pd.DataFrame:
```
Easy to follow the logic in this method!

ehanson8 commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,62 @@
+    def default_engine_parameters() -> dict:
+        return {"thick_mode": True}
+
+    def validate_data_warehouse_connection_string(self) -> None:
```
I think this implies some validation that doesn't appear to be happening, it seems to just check if it's set. I don't know if `verify_` is much better but I would at least update the docstring so it's not implying that it's checking the validity of the configuration

ghukill replied:
Nice catch.  I recall thinking this method might grow a bit, but better to keep it simple and accurate at the moment.  Renaming to `verify_connection_string_set()`.

ehanson8 commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,62 @@
+    engine_parameters: dict | None = field(default=None)
+    engine: Engine = field(default=None)
+
+    @staticmethod
+    def default_engine_parameters() -> dict:
+        return {"thick_mode": True}
```
Could this just be a class attribute?

ghukill replied:
Brilliant!  Agreed.  I twisted myself in a knot when I thought the parameters were going to be a mix-in approach, but it's either defaults or user provided.  This is much, much simpler:

```python
@define
class DWClient:
    """Client to provide Oracle Data Warehouse connection and querying."""

    connection_string: str = field(
        factory=lambda: Config().DATA_WAREHOUSE_CONNECTION_STRING,
        repr=False,
    )
    engine_parameters: dict = field(factory=lambda: {"thick_mode": True})   #<---------------------
    engine: Engine = field(default=None)
```

Which then requires zero logic in `init_engine()`.

ehanson8 commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,62 @@
+
+        User provided engine parameters will override self.default_engine_parameters.
+        """
+        self.validate_data_warehouse_connection_string()
+        if not self.engine:
+            engine_parameters = (
+                self.engine_parameters
+                if self.engine_parameters is not None
+                else self.default_engine_parameters()
+            )
+            self.engine = create_engine(self.connection_string, **engine_parameters)
```
Again, very clean and concise logic!

ghukill replied:
Thanks, but hilariously, not needed anymore!  Which is good, always nice to remove code where possible.

ehanson8 commented on tests/test_base_task.py:
```diff
@@ -143,3 +149,92 @@ def test_quickbase_task_run_upsert_and_json_receipt_output_target_success(
+):
+    foo_val, bar_val = 42, "apple"
+
+    class SQLQueryWithParameters(SQLQueryExtractTask):
+        stage = luigi.Parameter("Extract")
+
+        @property
+        def dwclient(self) -> DWClient:
+            return sqlite_dwclient
+
+        @property
+        def sql_query(self) -> str:
+            return """
+            select
+                :foo_val as foo,
+                :bar_val as bar
+            """
+
+        @property
+        def sql_query_parameters(self) -> dict:
+            return {"foo_val": foo_val, "bar_val": bar_val}
```
Could this be shifted to `tests/fixtures/tasks/extract.py` or `conftest.py`? This is large enough that defining it in the test feels messy

ghukill replied:
Yep, good call, don't see any reason why it can't live in `tests/fixtures/tasks/extract.py`.

ehanson8 reviewed (commented):
Seems solid, a few suggestions and questions

ghukill committed:
Code review updates

ehanson8 reviewed (approved):
Looks good but still recommend updating that docstring. Edit: don't know why this didn't link properly https://github.com/MITLibraries/hrqb-client/commit/16ad2a7c81ba71b19f39663328e0716852d725cc#r141964334

jonavellecuerdo commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,53 @@
+    def init_engine(self) -> None:
+        """Instantiate a SQLAlchemy engine if not already configured and set.
+
+        User provided engine parameters will override self.default_engine_parameters.
```
Should this be updated: `self.default_engine_parameters` -> `self.engine_parameters`?

ghukill replied:
Good catch!  Updated, and even better, can remove that altogether; I think setting `engine_parameters` explicitly is pretty transparent in how it works now.

jonavellecuerdo commented on hrqb/utils/data_warehouse.py:
```diff
@@ -0,0 +1,53 @@
+        factory=lambda: Config().DATA_WAREHOUSE_CONNECTION_STRING,
+        repr=False,
+    )
+    engine_parameters: dict = field(factory=lambda: {"thick_mode": True})
```
Would `field(default={"thick_mode": True})` be sufficient in this case as it doesn't call any methods to set the value? 🤔

ghukill replied:
Yep!  Given the `DATA_WAREHOUSE_CONNECTIONS_STRING` env var, these are the only SQLAlchemy engine parameters required (at least so far in testing).

ghukill replied:
Related to this comment and another, updated the docstring for this class to explain the fields:

```
"""Client to provide Oracle Data Warehouse connection and querying.

Fields:
    connection_string: str
        - full SQLAlchemy connection string, e.g.
            - oracle: oracle+oracledb://user1:pass1@example.org:1521/ABCDE
            - sqlite: sqlite:///:memory:
        - defaults to env var DATA_WAREHOUSE_CONNECTION_STRING, loaded from env vars
        at time of DWClient initialization
    engine_parameters: dict
        - optional dictionary of SQLAlchemy engine parameters
    engine: Engine
        - set via self.init_engine()
"""
```

jonavellecuerdo commented on hrqb/base/task.py:
```diff
@@ -125,6 +126,45 @@ def run(self) -> None:
+        if self.sql_query:
+            query = self.sql_query
+        elif self.sql_file:
+            with open(self.sql_file) as f:
```
Where would this SQL files live and in what cases is a SQL file preferred over a string? 🤔 

Hmm, what do you think of an update like:

```
class SQLQueryExtractTask(PandasPickleTask):
    """Base class for Tasks that make SQL queries for data."""

   ...

    @property
    def sql_query(self) -> str | None:
        """SQL query from string or loaded from file"""
        if self.sql_query:
            query = self.sql_query
        elif self.sql_file:
            with open(self.sql_file) as f:
                query = f.read()
        else:
            message = "Property sql_query or sql_file must be set."
            raise AttributeError(message)
        return query

    @property
    def sql_query_string(self) -> str | None:
        """SQL query from string to execute."""
        return None

    @property
    def sql_file(self) -> str | None:
        """SQL query loaded from file to execute."""
        return None

    @property
    def sql_query_parameters(self) -> dict:
        """Optional parameters to include with SQL query."""
        return {}

    def get_dataframe(self) -> pd.DataFrame:
        return self.dwclient.execute_query(
            self.sql_query,
            params=self.sql_query_parameters,
        )
```

ghukill replied:
The question and refactor suggestion are well received.  In fact, given some work since this PR on actually scoping out some extract tasks, it seems pretty clear that most (if not all) SQL extract tasks will use a file.

However, being able to easily override that with an explicit string is handy for testing.  I think there is a middleground here:

- `self.sql_query` returns a string, defaulting to reading `self.sql_file`
- if neither are overridden, throws an exception that `sql_file` must be defined OR `sql_query` overridden to provide a string vs reading a file

Going to make that change and  push a commit.

jonavellecuerdo reviewed (commented):
Looking good! Just have a few small suggestions / questions. 🤓

ghukill committed:
Additional DWClient docstrings

ghukill committed:
Default SQLQueryExtractTask to expect SQL file

Why these changes are being introduced:

During code review, and some work on sketching actual extract tasks,
it is clear that SQL extract tasks will most likely use SQL files.
However, it remains helpful for testing and other dev work to define
the query directly in the class.

How this addresses that need:
* Default SQLQueryExtractTask.sql_query to return string, and get
this string from SQLQueryExtractTask.sql_file
* Allows for overridding SQLQueryExtractTask.sql_query with an
explicit string

Side effects of this change:
* None

Relevant ticket(s):
* https://mitlibraries.atlassian.net/browse/HRQB-11

ghukill commented:
@jonavellecuerdo - added a couple of commits addressing your feedback.

jonavellecuerdo commented on hrqb/base/task.py:
```diff
@@ -135,9 +135,20 @@ def dwclient(self) -> DWClient:
+        """
+        if not self.sql_file:
+            message = (
+                "Property 'sql_file' must be set or property 'sql_query' overridden to "
```
Hmm, when we say that the "property X must set/overridden", does it mean: "the `@property` method must be defined for the class"? 

For testing purposes, it would be useful if we could easily set the values like so: 

`SQLQueryExtractTask().sql_file = "<filepath>"` or `SQLQueryExtractTask().sql_query = "<query>"`

I ran some test code and wasn't able to directly set these properties. 🤔 
```
class test:
    @property
    def prop(self):
        return None
```

Output:
```
>>> t.prop
>>> t.prop = "Hey!"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: property 'prop' of 'test' object has no setter
>>> 
```
---

In short, I think we need [`setter` methods for these `@property`-decorated methods if we want to be able to directly set the values](https://realpython.com/python-property/#using-property-as-a-decorator) _or_ update the logged message to clarify. Feels a bit tricky because it's like..an "optionally abstract"™️ method 😅 

Here's some test code:
```
class test:
    def __init__(self, prop=None) -> None:
        self._prop = prop

    @property
    def prop(self):
        return self._prop

    @prop.setter
    def prop(self, value):
        print(f"Setting to {value}")
        self._prop = value
```
Output: 
```
>>> t = test()
>>> t.prop
>>> t.prop = "cat"
Setting to cat
>>> t.prop
'cat'
```

ghukill replied:
Thanks for digging into this with me @jonavellecuerdo.

Counter-point: as a base Luigi task, these properties are designed to be defined / not-defined when we create _new_ task classes that extend this base class; not set directly.

Here is an example coming from a real extract task coming in future PRs:

```python
class ExtractDWEmployeeAppointments(SQLQueryExtractTask):
    """Query Data Warehouse for employee appointment data."""

    pipeline = luigi.Parameter()
    stage = luigi.Parameter("Extract")

    @property
    def sql_file(self) -> str:
        return "hrqb/tasks/sql/employee_appointments.sql"
```

Hoping this demonstrates that it's quite normal to override a property in this way, and not set it directly.

And [here is an example from the testing fixtures](https://github.com/MITLibraries/hrqb-client/pull/16/files#diff-52a8ed9fa9fb3bfc1b3996f5eb619d05ffd78905a66250ae543c4054865eaefaR49-R51).

I take your point that manually setting a property to override it is somewhat inconvenient in testing, but I think it's more analogous to overrding a method with default behavior via class inheritance.

And while it's _kind_ of like an abstract method, it technically does not need to be defined.  It would also be acceptable to just define `sql_query` and provide the SQL query explicitly, like [this test fixture demonstrates](https://github.com/MITLibraries/hrqb-client/pull/16/files#diff-52a8ed9fa9fb3bfc1b3996f5eb619d05ffd78905a66250ae543c4054865eaefaR65-R69).

Thoughts?  You okay with proceeding with this?

ghukill replied:
FWIW, I had considered making it a `luigi.Parameter()`, and I suppose it still could be, but I don't think that made sense for the `sql_query` parameter, and thought the symmetry was ultimately a better fit.

Ultimately, these base classes are for convenience when defining new classes, but I think it's implied (or it should be) that when defining classes, whatever works and makes sense in that instance is okay.

jonavellecuerdo replied:
@ghukill Ah, thanks for pointing out the test fixture! Yep, sounds good to me.

jonavellecuerdo reviewed (commented):
@ghukill Thank you for taking the time to make this update. Just one follow-up question!

jonavellecuerdo reviewed (commented):
~@ghukill Actually, can we still update the logged message~ Ignore!
